### ISSUE TITLE
Include clang-tidy in the list of packaged tools

### ISSUE THREAD

**Issue by HighCommander4** [elfshaker/manyclangs#9]
I notice `clang-tidy` is not included in the list of executables included in the package that are listed in https://github.com/elfshaker/manyclangs#more-on-linking.

Would it be possible to include it?

**Comment by pwaller** [comment]
Ah, I see this is not included because we [don't currently include the `clang-tools-extra` project in our build](https://github.com/elfshaker/elfshaker/blob/d8ec61253ef5ebc84de41d54a03b68f91e7dbad8/contrib/manyclangs-cmake#L53). Maybe one day, when we rebuild the packs, but it depends how much that blows up the cost of the build.

**Comment by HighCommander4** [comment]
Is there something I could do to help move this effort forward?

Having `clang-tools-extra` binaries such as `clang-tidy` and `clangd` in the archives would be super helpful for tracking down regressions in these tools.

**Comment by peterwaller-arm** [comment]
Sorry, I think this is blocked on me rebuilding the packs etc. Good to know you're interested though. Feel free to keep pinging every once in a while and I'll try and let you know if things change.

**Comment by peterwaller-arm** [comment]
Actually, it would be useful to know how far back you need. If you're happy with it just appearing in current/future packs then I can investigate turning it on for the next pack build and check that it isn't prohibitive. This might still take a moment though as it is low priority for me for the time being.

**Comment by HighCommander4** [comment]
> Actually, it would be useful to know how far back you need. If you're happy with it just appearing in current/future packs then I can investigate turning it on for the next pack build and check that it isn't prohibitive.

I typically use it for investigating regressions in the last release, or maybe last two releases. So, even if clang-tools-extra only started to appear in packs going forward, it's going to start being useful in a few months for tracking down clang-tidy and clangd regressions appearing in llvm-20.

**Comment by peterwaller-arm** [comment]
You are in luck, I've started adding it as of the 202407 pack forwards. Please can you give it a quick test on a few arbitrary commits to make sure it works as you need it to?

**Comment by peterwaller-arm** [comment]
Problem for another time: I had some issues I couldn't instantly resolve with clangd, so it is currently built with `CLANG_ENABLE_CLANGD=Off`. I think the issue is that the manyclangs build script avoids lots of I/O by supplying `-o /dev/null` to links unless those binaries are known needed for the build process; it is failing in a non-obvious way such that I couldn't quickly find binary I needed to exclude from this.

**Comment by HighCommander4** [comment]
> You are in luck, I've started adding it as of the 202407 pack forwards. Please can you give it a quick test on a few arbitrary commits to make sure it works as you need it to?

It seems to be working well -- thanks very much for adding!

**Comment by peterwaller-arm** [comment]
FYI @HighCommander4 -- elfshaker/elfshaker#150 fixed a bug which means commits which only modified clang-tools-extra would be omitted. From August 2025 onwards they'll be included.

**Comment by HighCommander4** [comment]
> FYI [@HighCommander4](https://github.com/HighCommander4) -- [elfshaker/elfshaker#150](https://github.com/elfshaker/elfshaker/pull/150) fixed a bug which means commits which only modified clang-tools-extra would be omitted. From August 2025 onwards they'll be included.

Thanks for the heads up and for fixing!

While we're on this topic, do you think it might be possible to revisit the idea of including clangd in the packages? Regarding the error you ran into previously:

> I think the issue is that the manyclangs build script avoids lots of I/O by supplying -o /dev/null to links unless those binaries are known needed for the build process; it is failing in a non-obvious way such that I couldn't quickly find binary I needed to exclude from this.

I'm happy to poke at this if you let me know what I need to do to get to the mentioned failure.

**Comment by peterwaller-arm** [comment]
If you want to have a go it's in contrib in the elfshaker repo.

If you get a blank Ubuntu machine you can do run the manyclangs-setup.sh:

https://github.com/elfshaker/elfshaker/blob/main/contrib/manyclangs-setup.sh
https://github.com/elfshaker/elfshaker/blob/3656bdefcc1d375f9a928b0fc566b16ff3362685/contrib/manyclangs-setup.sh#L241

Then the above command (invoking manyclangs-build-month with appropriate environment/args) will produce a month pack. It requires about 1 CPU-month of CPU time or on the order of magnitude, so scale according to your number of CPUs to understand the wall clock time required. Meanwhile you can inspect the log files found in /build/<date>/*.log to see what's happening under the hood in terms of the build. This is obviously a little wasteful but illustrates the process that I'm using, from the entry point.

If you want to avoid the whole pack you can try invoking https://github.com/elfshaker/elfshaker/blob/main/contrib/manyclangs-build directly but you'll need to figure out the prerequisite environment/working directory and arguments from manyclangs-build-month.

**Comment by HighCommander4** [comment]
> If you want to have a go it's in contrib in the elfshaker repo.
> 
> [...]

Thanks! I think I got it to work; PR at https://github.com/elfshaker/elfshaker/pull/152.
